File: web-sdk/src/data/repositories/yalo-message/sdk-message-mapper.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import {
  ChatMessage,
  type MessageButton,
  type MessageButtonType,
} from '@domain/models/chat-message/chat-message';
import {
  ButtonType,
  MessageRole,
  MessageStatus,
  ProductMessageRequest_Orientation,
  type Button as ProtoButton,
  type PollMessageItem,
  type Product as ProtoProduct,
  type SdkMessage,
} from '@domain/models/events/external_channel/in_app/sdk/sdk_message';
import { Product } from '@domain/models/product/product';

export function chatMessageToSdkMessage(
  message: ChatMessage,
  mediaId?: string
): SdkMessage {
  const timestamp = new Date();
  const correlationId = message.id?.toString() || '';

  switch (message.type) {
    case 'text':
      return {
        correlationId,
        textMessageRequest: {
          content: {
            timestamp: message.timestamp,
            text: message.content,
            status: MessageStatus.MESSAGE_STATUS_IN_PROGRESS,
            role: MessageRole.MESSAGE_ROLE_USER,
          },
          timestamp,
          buttons: [],
        },
        timestamp,
      };
    case 'image':
      return {
        correlationId,
        imageMessageRequest: {
          content: {
            timestamp: message.timestamp,
            text: message.content,
            status: MessageStatus.MESSAGE_STATUS_IN_PROGRESS,
            role: MessageRole.MESSAGE_ROLE_USER,
            mediaUrl: mediaId ?? message.fileName!,
            mediaType: message.mediaType!,
            byteCount: message.byteCount!,
            fileName: message.fileName!,
          },
          timestamp,
          buttons: [],
        },
        timestamp,
      };
    case 'voice':
      return {
        correlationId,
        voiceNoteMessageRequest: {
          content: {
            timestamp: message.timestamp,
            status: MessageStatus.MESSAGE_STATUS_IN_PROGRESS,
            role: MessageRole.MESSAGE_ROLE_USER,
            mediaUrl: mediaId ?? message.fileName!,
            mediaType: message.mediaType!,
            byteCount: message.byteCount!,
            fileName: message.fileName!,
            amplitudesPreview: message.amplitudes!,
            duration: message.duration!,
          },
          timestamp,
          buttons: [],
        },
        timestamp,
      };
    case 'video':
      return {
        correlationId,
        videoMessageRequest: {
          content: {
            timestamp: message.timestamp,
            text: message.content,
            status: MessageStatus.MESSAGE_STATUS_IN_PROGRESS,
            role: MessageRole.MESSAGE_ROLE_USER,
            mediaUrl: mediaId ?? message.fileName!,
            mediaType: message.mediaType!,
            byteCount: message.byteCount!,
            fileName: message.fileName!,
            duration: message.duration!,
          },
          timestamp,
          buttons: [],
        },
        timestamp,
      };
    case 'attachment':
      return {
        correlationId,
        attachmentMessageRequest: {
          content: {
            timestamp: message.timestamp,
            text: message.content,
            status: MessageStatus.MESSAGE_STATUS_IN_PROGRESS,
            role: MessageRole.MESSAGE_ROLE_USER,
            mediaUrl: mediaId ?? message.fileName!,
            mediaType: message.mediaType!,
            byteCount: message.byteCount!,
            fileName: message.fileName!,
          },
          timestamp,
          buttons: [],
        },
        timestamp,
      };
    default:
      throw Error('UnimplementedError');
  }
}

export function pollMessageItemToChatMessage(
  item: PollMessageItem
): ChatMessage | null {
  const timestamp = item.date ?? new Date();
  const msg = item.message;
  if (!msg) return null;

  if (msg.textMessageRequest?.content) {
    return ChatMessage.text({
      role: 'AGENT',
      timestamp,
      content: msg.textMessageRequest.content.text,
      wiId: item.id,
      header: msg.textMessageRequest.header,
      footer: msg.textMessageRequest.footer,
      buttons: toMessageButtons(msg.textMessageRequest.buttons),
    });
  }

  if (msg.imageMessageRequest?.content) {
    const content = msg.imageMessageRequest.content;
    return ChatMessage.image({
      role: 'AGENT',
      timestamp,
      fileName: content.mediaUrl || content.fileName,
      content: content.text ?? '',
      mediaType: content.mediaType,
      byteCount: content.byteCount,
      wiId: item.id,
      header: msg.imageMessageRequest.header,
      footer: msg.imageMessageRequest.footer,
      buttons: toMessageButtons(msg.imageMessageRequest.buttons),
    });
  }

  if (msg.voiceNoteMessageRequest?.content) {
    const content = msg.voiceNoteMessageRequest.content;
    return ChatMessage.voice({
      role: 'AGENT',
      timestamp,
      fileName: content.fileName,
      amplitudes: content.amplitudesPreview,
      duration: content.duration,
      mediaType: content.mediaType,
      byteCount: content.byteCount,
      wiId: item.id,
      header: msg.voiceNoteMessageRequest.header,
      footer: msg.voiceNoteMessageRequest.footer,
      buttons: toMessageButtons(msg.voiceNoteMessageRequest.buttons),
    });
  }

  if (msg.videoMessageRequest?.content) {
    const content = msg.videoMessageRequest.content;
    return ChatMessage.video({
      role: 'AGENT',
      timestamp,
      fileName: content.mediaUrl || content.fileName,
      content: content.text ?? '',
      duration: content.duration,
      mediaType: content.mediaType,
      byteCount: content.byteCount,
      wiId: item.id,
      header: msg.videoMessageRequest.header,
      footer: msg.videoMessageRequest.footer,
      buttons: toMessageButtons(msg.videoMessageRequest.buttons),
    });
  }

  if (msg.attachmentMessageRequest?.content) {
    const content = msg.attachmentMessageRequest.content;
    return ChatMessage.attachment({
      role: 'AGENT',
      timestamp,
      fileName: content.mediaUrl || content.fileName,
      content: content.text ?? '',
      mediaType: content.mediaType,
      byteCount: content.byteCount,
      wiId: item.id,
      header: msg.attachmentMessageRequest.header,
      footer: msg.attachmentMessageRequest.footer,
      buttons: toMessageButtons(msg.attachmentMessageRequest.buttons),
    });
  }

  if (msg.productMessageRequest) {
    const products = msg.productMessageRequest.products.map(toDomainProduct);
    const isCarousel =
      msg.productMessageRequest.orientation ===
      ProductMessageRequest_Orientation.ORIENTATION_HORIZONTAL;
    const factory = isCarousel ? ChatMessage.carousel : ChatMessage.product;
    return factory({
      role: 'AGENT',
      timestamp,
      products,
      wiId: item.id,
    });
  }

  return null;
}

function toMessageButtons(
  buttons: ProtoButton[] | undefined
): MessageButton[] {
  return buttons?.map(toMessageButton) ?? [];
}

function toMessageButton(b: ProtoButton): MessageButton {
  return { text: b.text, type: toMessageButtonType(b.buttonType), url: b.url };
}

function toMessageButtonType(t: ButtonType): MessageButtonType {
  switch (t) {
    case ButtonType.BUTTON_TYPE_LINK:
      return 'link';
    case ButtonType.BUTTON_TYPE_POSTBACK:
      return 'postback';
    case ButtonType.BUTTON_TYPE_REPLY:
    default:
      return 'reply';
  }
}

function toDomainProduct(p: ProtoProduct): Product {
  return new Product({
    sku: p.sku,
    name: p.name,
    price: p.price,
    imagesUrl: p.imagesUrl,
    salePrice: p.salePrice,
    subunits: p.subunits,
    unitStep: p.unitStep,
    unitName: p.unitName,
    subunitName: p.subunitName,
    subunitStep: p.subunitStep,
    unitsAdded: p.unitsAdded,
    subunitsAdded: p.subunitsAdded,
  });
}

File: flutter-sdk/lib/src/data/repositories/yalo_message/sdk_message_mapper.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'dart:io';
import 'dart:typed_data';

import 'package:fixnum/fixnum.dart';
import 'package:logging/logging.dart';
import 'package:mime/mime.dart';
import 'package:protobuf/well_known_types/google/protobuf/timestamp.pb.dart';
import 'package:uuid/uuid.dart';
import 'package:yalo_chat_flutter_sdk/domain/models/product/product.dart';
import 'package:yalo_chat_flutter_sdk/src/common/result.dart';
import 'package:yalo_chat_flutter_sdk/src/data/services/yalo_media/yalo_media_service.dart';
import 'package:yalo_chat_flutter_sdk/src/domain/models/chat_message/button.dart';
import 'package:yalo_chat_flutter_sdk/src/domain/models/chat_message/chat_message.dart';
import 'package:yalo_chat_flutter_sdk/src/domain/models/events/external_channel/in_app/sdk/sdk_message.pb.dart'
    as proto;
import 'package:yalo_chat_flutter_sdk/src/domain/models/events/external_channel/in_app/sdk/sdk_message.pbenum.dart'
    as proto_enum;

final _log = Logger('SdkMessageMapper');

Future<ChatMessage?> pollMessageItemToChatMessage(
  proto.PollMessageItem item, {
  required YaloMediaService mediaService,
  required Future<Directory> Function() directory,
}) async {
  switch (item.message.whichPayload()) {
    case proto.SdkMessage_Payload.textMessageRequest:
      final request = item.message.textMessageRequest;
      return ChatMessage.text(
        role: MessageRole.assistant,
        timestamp: item.date.toDateTime(),
        content: request.content.text,
        header: request.hasHeader() ? request.header : null,
        footer: request.hasFooter() ? request.footer : null,
        buttons: request.buttons.map(_toDomainButton).toList(),
        wiId: item.id,
      );
    case proto.SdkMessage_Payload.imageMessageRequest:
      final request = item.message.imageMessageRequest;
      final content = request.content;
      final downloadResult = await mediaService.downloadMedia(content.mediaUrl);
      switch (downloadResult) {
        case Ok(:final Uint8List result):
          final mimeType = content.mediaType.isNotEmpty
              ? content.mediaType
              : (lookupMimeType(content.mediaUrl) ?? 'image/jpeg');
          final ext = extensionFromMime(mimeType) ?? 'jpg';
          final dir = await directory();
          final localPath = '${dir.path}/${const Uuid().v4()}.$ext';
          await File(localPath).writeAsBytes(result);
          return ChatMessage.image(
            role: MessageRole.assistant,
            timestamp: item.date.toDateTime(),
            content: content.text,
            fileName: localPath,
            mediaType: mimeType,
            byteCount: result.length,
            header: request.hasHeader() ? request.header : null,
            footer: request.hasFooter() ? request.footer : null,
            buttons: request.buttons.map(_toDomainButton).toList(),
            wiId: item.id,
          );
        case Error(:final error):
          _log.severe('Failed to download image for message ${item.id}', error);
          return null;
      }
    case proto.SdkMessage_Payload.videoMessageRequest:
      final request = item.message.videoMessageRequest;
      final content = request.content;
      final downloadResult = await mediaService.downloadMedia(content.mediaUrl);
      switch (downloadResult) {
        case Ok(:final Uint8List result):
          final mimeType = content.mediaType.isNotEmpty
              ? content.mediaType
              : (lookupMimeType(content.mediaUrl) ?? 'video/mp4');
          final ext = extensionFromMime(mimeType) ?? 'mp4';
          final dir = await directory();
          final localPath = '${dir.path}/${const Uuid().v4()}.$ext';
          await File(localPath).writeAsBytes(result);
          return ChatMessage.video(
            role: MessageRole.assistant,
            timestamp: item.date.toDateTime(),
            content: content.text,
            fileName: localPath,
            duration: content.duration.toInt(),
            mediaType: mimeType,
            byteCount: result.length,
            header: request.hasHeader() ? request.header : null,
            footer: request.hasFooter() ? request.footer : null,
            buttons: request.buttons.map(_toDomainButton).toList(),
            wiId: item.id,
          );
        case Error(:final error):
          _log.severe('Failed to download video for message ${item.id}', error);
          return null;
      }
    case proto.SdkMessage_Payload.productMessageRequest:
      final List<Product> products = item.message.productMessageRequest.products
          .map(_toDomainProduct)
          .toList();
      final bool isCarousel =
          item.message.productMessageRequest.orientation ==
          proto_enum
              .ProductMessageRequest_Orientation
              .ORIENTATION_HORIZONTAL;
      if (isCarousel) {
        return ChatMessage.carousel(
          role: MessageRole.assistant,
          timestamp: item.date.toDateTime(),
          products: products,
          wiId: item.id,
        );
      }
      return ChatMessage.product(
        role: MessageRole.assistant,
        timestamp: item.date.toDateTime(),
        products: products,
        wiId: item.id,
      );
    case _:
      return null;
  }
}

proto.SdkMessage? chatMessageToSdkMessage(
  ChatMessage chatMessage, {
  String? mediaId,
}) {
  final messageStatus = switch (chatMessage.status) {
    MessageStatus.delivered => proto.MessageStatus.MESSAGE_STATUS_DELIVERED,
    MessageStatus.read => proto.MessageStatus.MESSAGE_STATUS_READ,
    MessageStatus.error => proto.MessageStatus.MESSAGE_STATUS_ERROR,
    MessageStatus.sent => proto.MessageStatus.MESSAGE_STATUS_SENT,
    MessageStatus.inProgress => proto.MessageStatus.MESSAGE_STATUS_IN_PROGRESS,
  };
  final timestamp = DateTime.now();

  final protoButtons = chatMessage.buttons.map(_toProtoButton).toList();

  return switch (chatMessage.type) {
    MessageType.text => proto.SdkMessage(
      correlationId: chatMessage.id.toString(),
      timestamp: Timestamp.fromDateTime(timestamp),
      textMessageRequest: proto.TextMessageRequest(
        content: proto.TextMessage(
          timestamp: Timestamp.fromDateTime(chatMessage.timestamp),
          text: chatMessage.content,
          status: messageStatus,
          role: proto.MessageRole.MESSAGE_ROLE_USER,
        ),
        header: chatMessage.header,
        footer: chatMessage.footer,
        buttons: protoButtons,
      ),
    ),
    MessageType.image => proto.SdkMessage(
      correlationId: chatMessage.id.toString(),
      timestamp: Timestamp.fromDateTime(timestamp),
      imageMessageRequest: proto.ImageMessageRequest(
        content: proto.ImageMessage(
          timestamp: Timestamp.fromDateTime(chatMessage.timestamp),
          text: chatMessage.content,
          fileName: chatMessage.fileName,
          mediaUrl: mediaId!,
          mediaType: chatMessage.mediaType,
          byteCount: Int64(chatMessage.byteCount!),
          status: messageStatus,
          role: proto.MessageRole.MESSAGE_ROLE_USER,
        ),
        header: chatMessage.header,
        footer: chatMessage.footer,
        buttons: protoButtons,
      ),
    ),
    MessageType.voice => proto.SdkMessage(
      correlationId: chatMessage.id.toString(),
      timestamp: Timestamp.fromDateTime(timestamp),
      voiceNoteMessageRequest: proto.VoiceNoteMessageRequest(
        content: proto.VoiceMessage(
          timestamp: Timestamp.fromDateTime(chatMessage.timestamp),
          fileName: chatMessage.fileName,
          mediaUrl: mediaId!,
          amplitudesPreview: chatMessage.amplitudes,
          duration: chatMessage.duration?.toDouble(),
          mediaType: chatMessage.mediaType,
          byteCount: Int64(chatMessage.byteCount!),
          status: messageStatus,
          role: proto.MessageRole.MESSAGE_ROLE_USER,
        ),
        header: chatMessage.header,
        footer: chatMessage.footer,
        buttons: protoButtons,
      ),
    ),
    MessageType.video => proto.SdkMessage(
      correlationId: chatMessage.id.toString(),
      timestamp: Timestamp.fromDateTime(timestamp),
      videoMessageRequest: proto.VideoMessageRequest(
        content: proto.VideoMessage(
          timestamp: Timestamp.fromDateTime(chatMessage.timestamp),
          text: chatMessage.content,
          fileName: chatMessage.fileName,
          mediaUrl: mediaId!,
          duration: chatMessage.duration?.toDouble(),
          mediaType: chatMessage.mediaType,
          byteCount: Int64(chatMessage.byteCount!),
          status: messageStatus,
          role: proto.MessageRole.MESSAGE_ROLE_USER,
        ),
        header: chatMessage.header,
        footer: chatMessage.footer,
        buttons: protoButtons,
      ),
    ),
    MessageType.product => null,
    MessageType.productCarousel => null,
    MessageType.promotion => null,
    MessageType.unknown => null,
  };
}

Button _toDomainButton(proto.Button b) {
  final ButtonType type;
  if (b.buttonType == proto_enum.ButtonType.BUTTON_TYPE_LINK) {
    type = ButtonType.link;
  } else if (b.buttonType == proto_enum.ButtonType.BUTTON_TYPE_POSTBACK) {
    type = ButtonType.postback;
  } else {
    type = ButtonType.reply;
  }
  return Button(
    text: b.text,
    type: type,
    url: b.hasUrl() ? b.url : null,
  );
}

proto.Button _toProtoButton(Button b) {
  return proto.Button(
    text: b.text,
    buttonType: switch (b.type) {
      ButtonType.reply => proto_enum.ButtonType.BUTTON_TYPE_REPLY,
      ButtonType.postback => proto_enum.ButtonType.BUTTON_TYPE_POSTBACK,
      ButtonType.link => proto_enum.ButtonType.BUTTON_TYPE_LINK,
    },
    url: b.url,
  );
}

Product _toDomainProduct(proto.Product p) {
  return Product(
    sku: p.sku,
    name: p.name,
    price: p.price,
    imagesUrl: p.imagesUrl.toList(),
    salePrice: p.hasSalePrice() ? p.salePrice : null,
    subunits: p.subunits,
    unitStep: p.unitStep,
    unitName: p.unitName,
    subunitName: p.hasSubunitName() ? p.subunitName : null,
    subunitStep: p.subunitStep,
    unitsAdded: p.unitsAdded,
    subunitsAdded: p.subunitsAdded,
  );
}
